core/src/main/java/org/elasticsearch/search/aggregations/pipeline/bucketsort/BucketSortPipelineAggregator.java

```diff
@@ -0,0 +1,194 @@
+                                        Map<String, Object> metadata) {
+        super(name, sorts.stream().map(s -> s.getFieldName()).toArray(String[]::new), metadata);
+        this.sorts = sorts;
+        this.from = from;
+        this.size = size;
+        this.gapPolicy = gapPolicy;
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public BucketSortPipelineAggregator(StreamInput in) throws IOException {
+        super(in);
+        sorts = in.readList(FieldSortBuilder::new);
+        from = in.readVInt();
+        size = in.readOptionalVInt();
+        gapPolicy = GapPolicy.readFrom(in);
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeList(sorts);
+        out.writeVInt(from);
+        out.writeOptionalVInt(size);
+        gapPolicy.writeTo(out);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return BucketSortPipelineAggregationBuilder.NAME;
+    }
+
+    @Override
+    public InternalAggregation reduce(InternalAggregation aggregation, ReduceContext reduceContext) {
+        InternalMultiBucketAggregation<InternalMultiBucketAggregation, InternalMultiBucketAggregation.InternalBucket> originalAgg =
+                (InternalMultiBucketAggregation<InternalMultiBucketAggregation, InternalMultiBucketAggregation.InternalBucket>) aggregation;
+        List<? extends InternalMultiBucketAggregation.InternalBucket> buckets = originalAgg.getBuckets();
+        int bucketsCount = buckets.size();
+        int offset = reduceContext.isFinalReduce() ? from : 0;
+        int currentSize = size == null ? bucketsCount : size;
+
+        if (offset >= bucketsCount) {
+            return originalAgg.create(Collections.emptyList());
+        }
+
+        int resultSize = Math.min(currentSize, bucketsCount - offset);
+
+        // If no sorting needs to take place, we just truncate and return
+        if (sorts.size() == 0) {
+            return originalAgg.create(truncate(buckets, offset, currentSize));
+        }
+
+        int queueSize = Math.min(offset + currentSize, bucketsCount);
+        PriorityQueue<ComparableBucket> ordered = new TopNPriorityQueue(queueSize);
+        for (InternalMultiBucketAggregation.InternalBucket bucket : buckets) {
+            ordered.insertWithOverflow(new ComparableBucket(originalAgg, bucket));
+        }
+
+        // Popping from the priority queue returns the least element. The elements we want to skip due to offset would pop last.
+        // Thus, we just have to pop as many elements as we expect in results and store them in reverse order.
+        LinkedList<InternalMultiBucketAggregation.InternalBucket> newBuckets = new LinkedList<>();
+        for (int i = resultSize - 1; i >= 0; --i) {
+            ComparableBucket comparableBucket = ordered.pop();
+            if (comparableBucket.skip == false) {
```

Comment:
I may be reading this wrong, but if a bucket is skipped I think the final list will be incorrectly sized?  E.g.

- `resultSize == size == 5`
- `bucketCount == 10`
- Bucket 2 is skipped

The conditional will prevent Bucket 2 from being added to the list, but the loop will decrement regardless and it'll only add four values to `newBuckets`, rather than grabbing Bucket 6 for a total of five results.

Would it be easier to just resolve the bucket value up front and if it's skipped, avoid adding it to the priority queue entirely?   Also see the note below in `ComparableBucket` regarding resolving the values